feat(test-validation): extend validation script to enforce test structure conventions

aliersh opened the pull request:
This PR expands the existing test name validation logic by introducing comprehensive structure validation rules for Forge test artifacts. The goal is to enforce consistency across test file organization, contract naming, and alignment with source contracts, as part of the ongoing test suite refactor.

## New Validations Introduced

### Test File Path Validation
- Ensures test contracts are located under `test/` and use the `.t.sol` suffix
- Requires a corresponding source file under `src/` with a matching path and `.sol` suffix

### Contract Name vs File Name
- Verifies that the contract name in the compilation target matches the test file name (e.g., `Foo.t.sol` → `Foo_*`)
- Rejects test contracts where the base name does not align with the file name

### Contract Naming Pattern Validation
Validates that test contract names follow one of the following patterns:
- `<Contract>_TestInit`
- `<Contract>_<Function>_Test` (e.g., `MyContract_DoSomething_Test`)
- `<Contract>_Uncategorized_Test` or `_Unclassified_Test`
- `<Contract>_<Descriptor>_Harness` or - `<Contract>_Harness` (e.g., `Foo_Upgrade_Harness`)

If pattern is `<Contract>_<Function>_Test`, it also checks that the function actually exists in the source contract ABI.

### Function Existence Validation
- Uses source artifacts to verify that any function mentioned in a test contract name (e.g., `Foo_Bar_Test`) exists in the source contract `Foo.sol`
- Supports special functions like `constructor`, `receive`, and `fallback`

### Exclusion Support
- Adds centralized exclusion list for files and directories that are non-standard, legacy, or vendor-related
- Ensures only relevant test artifacts are validated

## Additional Changes
- Modularized the validation logic into functions for better structure and readability
- Added detailed documentation to each validation function
- Renamed the validation script directory from `test-names` to `test-validation`
- Updated `justfile` commands to reflect the new script location and responsibilities

---

This lays the groundwork for enforcing project-wide consistency and simplifies future onboarding, review, and maintenance.

smartcontracts commented on packages/contracts-bedrock/test/L1/SystemConfig.t.sol:
```diff
@@ -313,9 +313,9 @@ contract SystemConfig_Upgrade_Test is SystemConfig_TestInit {
     }
 }
 
-/// @title SystemConfig_StartBlock_TestFail
```
I recommend splitting apart these changes and fixing these issues in a separate PR. I would just "ignore" these files in the script for now, which will keep the PR much smaller. Later you can open up separate PRs to do the fixing, which will clear up the list of ignored files.

aliersh replied:
done in [#cd1601f](https://github.com/ethereum-optimism/optimism/pull/16543/commits/cd1601f88ab0befa8b1c423fde22247ec2e207a9)

aliersh committed:
chore(test-validation): add full test-artifact validation framework

* Introduces `scripts/checks/test-validation/` (migrated from `test-names/`)
  * `main.go`, helpers, and unit tests
* **New rules**
  * **Path rule** – test files must live under `test/**` and mirror a source file under `src/**`
  * **Contract-vs-file rule** – contract name in the compilation target must match the test file basename (`Foo.t.sol` → `Foo_*`)
  * **Naming-pattern rule** – allows only:
    * `<Contract>_TestInit`
    * `<Contract>_<Function>_Test` (function must exist in ABI)
    * `<Contract>_Uncategorized_Test` / `_Unclassified_Test`
    * `<Contract>_<Descriptor>_Harness` or `<Contract>_Harness`
  * **Function-existence rule** – verifies constructor / receive / fallback / selector names where required
* Adds central **exclusion list** for legacy or vendor test paths so the validator can ship without failing CI
* Refactors code into small, documented helpers (`validateTestName`, `validateTestStructure`, `checkSrcPath`, etc.)
* Updates `justfile` with `just test-validation` target
* Provides extensive inline docs to ease future rule additions

aliersh committed:
docs(test-validation): add detailed comments explaining exclusion reasons

- Add comprehensive documentation for each exclusion category
- Explain SRC validation exclusions for non-standard naming conventions
- Document contract name validation exclusions for mismatched names
- Clarify function name validation exclusions for lookup failures
- Describe test pattern validation exclusions for non-standard patterns
- Include inline comments for each excluded path with specific reasons
